parquet-hadoop/src/main/java/org/apache/parquet/crypto/CryptoPropertiesFactory.java

```diff
@@ -51,20 +71,36 @@ public static CryptoPropertiesFactory loadFactory(Configuration conf) throws IOE
 
     try {
       CryptoPropertiesFactory cryptoFactory = (CryptoPropertiesFactory)cryptoPropertiesFactoryClass.newInstance();
-      cryptoFactory.initialize(conf);
       return cryptoFactory;
     } catch (InstantiationException | IllegalAccessException e) {
       throw new BadConfigurationException("could not instantiate CryptoPropertiesFactory class: "
         + cryptoPropertiesFactoryClass, e);
     }
   }
 
-  public abstract void initialize(Configuration hadoopConfig);
-
-  public abstract FileEncryptionProperties getFileEncryptionProperties(
-    Configuration fileHadoopConfig, Path tempFilePath,
-    WriteContext fileWriteContext)  throws IOException;
+  /**
+   * Get FileEncryptionProperties object which is created by the implementation of this interface. Please see
+   * the unit test (TBD) for example
+   *
+   * @param fileHadoopConfig Configuration that is used to pass the needed information, e.g. KMS uri
+   * @param tempFilePath File path of the parquet file being written.
+   *                     Can be used for AAD prefix creation, key material management, etc.
+   *                     Implementations must not presume the path is permanent,
+   *                     as the file can be moved or renamed later
+   * @param fileWriteContext WriteContext to provide information like schema to build the FileEncryptionProperties
+   * @return
+   * @throws IOException
+   */
+  FileEncryptionProperties getFileEncryptionProperties(Configuration fileHadoopConfig, Path tempFilePath,
+                                                       WriteContext fileWriteContext)  throws IOException;
```

Comment:
Why do we think that an `IOException` might occur? Would a file read/write suppose to happen in an implementation?

Reply:
In our implementation, a number of different things can go wrong when creating encryption (decryption) properties, ranging from a failure to communicate with a KMS system, to having a wrong access control token, to missing the required hadoop properties (such as a list of encrypted columns), to failure to write/read separate files that keep the key material. Some of these problems belong to the IO domain, some don't - but my impression was that most of Parquet API exceptions are the IOException, because Parquet is an IO library (file writing/reading).

Reply:
There is always a question of what the application (that calls Parquet API) can do differently, in an automatic fashion, for different exception types. In this case, I can't see such automatic differentiated response. If the creation of encr/decr properties fails, the only thing to do is to see exception message, and either fix the KMS comms (restart, fix network, ), or provide a valid access token (often manual procedure), or to configure the required parameters, or to fix the filesystem problems, etc.

Reply:
I see.
What I don't like in IOException in this case is that it is a checked exception so when we call these methods we have to handle them or let them thrown out of the library. I don't think we can do anything useful at handling so we have to throw an exception. We can either let through the IOException in the call stack which can be very painful or even impossible if we don't want to break the existing API or we can wrap the IOException into a runtime exception. 
Usually, the latter is the preferred way in parquet-mr. So why would we expect an IOException from our internal API (even thought it is implemented by our clients) if we will wrap it into our own runtime exception anyway.
I would suggest expecting a `ParquetEncodingException` or a `ParquetDecodingException` or even create specific ones for this case (by extending a `ParquetRuntimeException`).

Reply:
Sounds good. I'm fine with either option.

Reply:
I will create ParquetCryptoRuntimeException extends ParquetRuntimeException